File: chisel/ui/macosx/label.m
#include <Cocoa/Cocoa.h>
#include <assert.h>

#include <chisel-native.h>
#include <chisel-native-ui.h>

#include <chisel-native-view.h>
#include <chisel-native-label.h>

native_handle _chisel_native_label_create( ) {
	NSTextField *view = [[NSTextField alloc] init];
	assert( view != nil );
	
	[view setEditable:NO];
	[view setBezeled:NO];
	[view setDrawsBackground:NO];
	
	return (native_handle)view;
}

void _chisel_native_label_set_selectable( native_handle native, int val ) {
	NSTextField *label = (NSTextField *)native;
	
	[label setSelectable: val == 0 ? NO : YES];
}

int _chisel_native_label_get_selectable( native_handle native ) {
	NSTextField *label = (NSTextField *)native;
	
	return [label isSelectable] == YES ? 1 : 0;
}

void _chisel_native_label_set_text( native_handle native, native_handle s ) {
	NSTextField *label = (NSTextField *)native;
	NSString *str = (NSString *)s;
	
	[label setStringValue: str];
}

CLFloat _chisel_native_label_get_height( native_handle native ) {
	NSTextField *label = (NSTextField *)native;
	NSTextFieldCell *cell = [label cell];
	NSSize cellSize = [cell cellSize];
	
	return cellSize.height;
}

File: chisel/ui/gtk/label.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include <gtk/gtk.h>

#include <chisel-native.h>
#include <chisel-native-ui.h>

#include <chisel-native-view.h>
#include <chisel-native-label.h>

native_handle _chisel_native_label_create( ) {
	GtkWidget *widget = gtk_label_new( "" );
	
	gtk_misc_set_alignment( GTK_MISC(widget), 0, 0 );
	
	return (native_handle)widget;
}

void _chisel_native_label_set_selectable( native_handle native, int val ) {
	gtk_label_set_selectable( GTK_LABEL(native), val );
}

int _chisel_native_label_get_selectable( native_handle native ) {
	return gtk_label_get_selectable( GTK_LABEL(native) );
}

void _chisel_native_label_set_text( native_handle native, native_handle s ) {
	GString *str = (GString *)s;
	
	gtk_label_set_text( GTK_LABEL(native), str->str );
}

CLFloat _chisel_native_label_get_height( native_handle native ) {
	return 16.0; // FIXME: hax
}